### ISSUE TITLE
alloca(3) usage warning

### ISSUE THREAD

**Issue by krytarowski** [dotnet/coreclr#3180]
```
../../../../../src/libcoreclrpal.a(path.cpp.o): In function `GetFullPathNameW':
/tmp/pkgsrc-tmp/wip/coreclr-git/work/coreclr/src/pal/src/file/path.cpp:227:
warning: Warning: reference to the libc supplied alloca(3); this most likely
will not work. Please use the compiler provided version of alloca(3), by
supplying the appropriate compiler flags (e.g. not -std=c89).
```

Part of `/usr/include/stdlib.h`

```
/*
 * Implementation-defined extensions
 */
#if defined(_NETBSD_SOURCE)
#if defined(alloca) && (alloca == __builtin_alloca) && \
        defined(__GNUC__) && (__GNUC__ < 2)
void    *alloca(int);     /* built-in for gcc */
#elif defined(__PCC__) && !defined(__GNUC__)
#define alloca(size) __builtin_alloca(size)
#else
void    *alloca(size_t);
#endif /* __GNUC__ */
```

```
$ pkg_info |grep -E 'lldb|llvm|clang|libunwind'                                                                                                     
llvm-3.9.0nb20160213 Low Level Virtual Machine compiler infrastructure
clang-3.9.0nb20160213 C language family frontend for LLVM
libunwind-3.9.0nb20160213 LLVM libunwind
lldb-3.9.0nb20160213 next generation, high-performance debugger
```

```
$ uname -a                                      
NetBSD chieftec 7.99.26 NetBSD 7.99.26 (GENERIC) #0: Wed Feb 10 21:58:18 UTC 2016  root@chieftec:/tmp/netbsd-tmp/sys/arch/amd64/compile/GENERIC amd64
```

**Comment by krytarowski** [comment]
@jkotas have you got an idea how to address it?

**Comment by krytarowski** [comment]
```
CopyFileA(
      IN LPCSTR lpExistingFileName,
      IN LPCSTR lpNewFileName,
      IN BOOL bFailIfExists)
{
    CPalThread *pThread;
    HANDLE       hSource = INVALID_HANDLE_VALUE;
    HANDLE       hDest = INVALID_HANDLE_VALUE;
    DWORD        dwDestCreationMode;
    BOOL         bGood = FALSE;
    DWORD        dwSrcFileAttributes;
    struct stat  SrcFileStats;

    LPSTR lpUnixPath = NULL;
    const int    buffer_size = 16*1024;
#if 0
    char        *buffer = (char*)alloca(buffer_size);
#else
    char         buffer[buffer_size];
#endif
```

The current usage `alloca`(3) is damaging stack and due to it tests are failing. For example the above diff code fixes:
- paltest_copyfilea_test1
- paltest_copyfilea_test2
- paltest_copyfilea_test3
- paltest_copyfilea_test4

**Comment by krytarowski** [comment]
I roughtly iterated over other failures and they are due to stack corruption as well.

**Comment by krytarowski** [comment]
Replacing `alloca`(3) right now with `VLA` isn't trivial - as there are like 300 uses in the code base.

**Comment by krytarowski** [comment]
I will try to define `alloca` as `PAL_alloca`

**Comment by jkotas** [comment]
I think `alloca` needs to get defined as `__builtin_alloca` - either directly or indirectly - to make things work.

**Comment by krytarowski** [comment]
I'm running out of ideas. I will try to rename all instances of `alloca(` to `_alloca(` -- and define `_alloca(` for !`VistualStudio` as `__builtin_alloca`.

**Comment by janvorli** [comment]
How about just definining alloca as __buildint_alloca for non-windows? Then you would not have to rename the alloca everywhere.

**Comment by coypoop** [comment]
krytarowski did it (trying to define explicity alloca and _alloca as __builtin_alloca), but it resulted in the same issue with alloca.

**Comment by krytarowski** [comment]
```
$ cat test.cpp
#include <cstdlib>
int main(){void *p = alloca(10);}
```

```
$ clang++ test.cpp -fno-builtin 
/var/tmp/test-7dc7d6.o: In function `main':
test.cpp:(.text+0x10): warning: Warning: reference to the libc supplied alloca(3); this most likely will not work. Please use the compiler provided version of alloca(3), by supplying the appropriate compiler flags (e.g. not -std=c89).
```

**Comment by krytarowski** [comment]
Fixed in #3189
